### PR TITLE
WriteFailures: Add metrics to increase monitoring

### PR DESCRIPTION
**Pull request by DylanGuedes**
**What this PR does / why we need it**:
Add metrics for our write failures manager. This way we can easily monitor how much logging we're doing and how much we're discarding.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: pkg/distributor/writefailures/metrics.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type metrics struct {
+	loggedCount    *prometheus.CounterVec
+	discardedCount *prometheus.CounterVec
+}
+
+func newMetrics(reg prometheus.Registerer, subsystem string) *metrics {
+	// prometheus.NewCounterVec(opts prometheus.CounterOpts, labelNames []string)
```
Comment (chaudum) [inline pkg/distributor/writefailures/metrics.go:14]: Can be removed.

File: pkg/distributor/writefailures/manager.go [hunk 2]
```diff
@@ -42,6 +45,10 @@ func (m *Manager) Log(tenantID string, err error) {
 
 	errMsg := err.Error()
 	if m.limiter.AllowN(time.Now(), tenantID, len(errMsg)) {
+		m.m.loggedCount.WithLabelValues(tenantID)
```
Comment (chaudum) [inline pkg/distributor/writefailures/manager.go:48]:
```suggestion
		m.m.loggedCount.WithLabelValues(tenantID).Inc()
```

File: pkg/distributor/writefailures/manager.go [hunk 3]
```diff
@@ -42,6 +45,10 @@ func (m *Manager) Log(tenantID string, err error) {
...
 		level.Error(m.logger).Log("msg", "write operation failed", "details", errMsg, "tenant", tenantID)
+		return
 	}
+
+	m.m.discardedCount.WithLabelValues(tenantID)
```
Comment (chaudum) [inline pkg/distributor/writefailures/manager.go:53]:
```suggestion
	m.m.discardedCount.WithLabelValues(tenantID).Inc()
```

File: pkg/distributor/writefailures/metrics.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	return &metrics{
+		loggedCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Namespace:   "loki",
+			Name:        "write_failures_logged_total",
+			Help:        "The total number log failures were logged for a tenant.",
```
Comment (chaudum) [inline pkg/distributor/writefailures/metrics.go:19]: This help text does not make sense to me.

**Review by chaudum — CHANGES_REQUESTED** [review]

File: pkg/distributor/writefailures/metrics.go (hunk 1 above)
Comment (DylanGuedes) [reply pkg/distributor/writefailures/metrics.go]: thank you.

File: pkg/distributor/writefailures/metrics.go (hunk 4 above)
Comment (DylanGuedes) [reply pkg/distributor/writefailures/metrics.go]: I pushed a comment rewriting it a little bit, lmk if it looks better.

File: pkg/distributor/writefailures/manager.go (hunk 2 above)
Comment (DylanGuedes) [reply pkg/distributor/writefailures/manager.go]: my brain was dying i'm pretty sure

File: pkg/distributor/writefailures/manager.go [hunk 5]
```diff
@@ -28,6 +30,7 @@ func NewManager(logger log.Logger, cfg Cfg, tenants *runtime.TenantConfigs) *Man
 		limiter:    limiter.NewRateLimiter(strategy, time.Minute),
 		logger:     logger,
 		tenantCfgs: tenants,
+		m:          newMetrics(reg, subsystem),
```
Comment (chaudum) [inline pkg/distributor/writefailures/manager.go:33]:
```suggestion
		metrics:    newMetrics(reg, subsystem),
```

Comment (DylanGuedes) [reply pkg/distributor/writefailures/manager.go]: fixed!

**Review by chaudum — APPROVED** [review]

**Pull request merged**